### PR TITLE
feat(ffi)!: refactor dataclasses, enums, and ObjectPtr containers

### PR DESCRIPTION
**Pull request by junrushao**
This PR is split into:
- https://github.com/apache/tvm-ffi/pull/663
- https://github.com/apache/tvm-ffi/pull/664
- https://github.com/apache/tvm-ffi/pull/665
- https://github.com/apache/tvm-ffi/pull/666

## Summary

- Refactor Python and native dataclass field resolution, conversion, and object-valued field lifetimes.
- Unify enum registration and lookup around canonical native singleton state.
- Add direct `ObjectPtr<TObject>` FFI traits for `Any`, reflection, serialization, and explicit typed storage.
- Redirect unqualified object template arguments such as `Map<TObject, V>` to `Map<ObjectPtr<TObject>, V>`.

## Architecture

- Resolve annotations against the class that declares each field, centralize typed conversion in `TypeAttr`/`ObjectDef`, and explicitly construct and destroy native object-valued fields.
- Store enum entries, canonical integer/string indexes, and extensible attributes in one `EnumState`; reuse the dataclass conversion path for payloads while preserving singleton identity.
- Teach the core runtime to retain, move, cast, reflect, traverse, and serialize `ObjectPtr<TObject>` directly.
- Normalize object template arguments through `object_ptr_type_t` only at the final container layer, keeping direct pointer support independently usable and preserving C++17 class-template deduction.

## Public Interfaces

- Add `init_property`, reflection conversion hooks, and owner-local dataclass field resolution.
- Make `@py_class` default to `structural_eq="tree"`; explicit `None` still opts out.
- Add native `EnumState`, `IntEnum`, and `StrEnum` types with canonical integer/string lookup.
- Allow unqualified `ObjectPtr<TObject>` values in `Any`, packed functions, reflected fields, and explicitly typed containers.
- Make `Array`, `List`, `Map`, `Dict`, `Tuple`, `Optional`, `Variant`, and `Expected` normalize unqualified `Object` subclasses to `ObjectPtr` storage.
- Keep shorthand and explicit pointer container spellings interoperable without copying backing storage.
- The stable TVM FFI C ABI is unchanged.

## UI/UX

- None.

## Behavioral Changes

- Forward references, recursive classes, and mixins resolve fields from their declaring owner, with full nested paths in conversion errors.
- `init_property` values are computed after FFI-backed initialization and excluded from constructor inputs and structural identity.
- Enum copy and JSON graph round trips restore registered singleton objects.
- Copying an `ObjectPtr` through `Any` retains it; moving transfers ownership; null maps to the ABI `None` value.
- Container accessors, schemas, checks, and ownership use `ObjectPtr<TObject>` for unqualified object template arguments. Qualified object types remain unsupported by the shorthand.

## Breaking Changes and Migration

- Pass `structural_eq=None` where code relies on the former `@py_class` opt-out default.
- Migrate native enum code from `_value`, `_name`, `Get`, and the old entry/attribute columns to `_int_index`, `_str_index`, typed index lookup, and `EnumState`.
- Treat unqualified object container arguments as pointer storage, or spell the intended storage type explicitly where exact declared-type behavior matters.
- Rebuild C++ and Cython extensions after migrating reflected declarations and enum/container code.

## Docs

- Document tree-by-default dataclass structural equality and enum singleton identity.
- Document enum payload serialization and canonical singleton restoration.
- Document direct `ObjectPtr` ownership, qualification/null limitations, and shorthand for every supported container family.

## Tests

- Every semantic boundary received a clean editable C++/Cython build.
- Dataclass boundary: 1103 focused Python tests passed; 2 expected failures remained expected.
- Enum boundary: 599 focused Python tests passed; a fresh Debug C++ test build passed; 4 targeted enum/serialization CTests passed.
- Direct `ObjectPtr` boundary: a clean Debug test build passed; all 19 `ObjectPtr`-matching CTests and 450 focused Python tests passed before container redirection was present.
- Final container boundary: all 444 enabled CTests passed; 2 existing cycle-detection tests remained disabled; 552 focused Python tests passed.
- Stub generation was idempotent on the complete tree, and scoped pre-commit hooks passed at every boundary.
- The final tree and stable aggregate patch ID exactly match the original squashed commit; each intermediate tree was independently audited for semantic leakage.
- The identical final tree had passed targeted clang-tidy before the history rewrite.

## Untested Edge Cases

- Rust tests, the remainder of the Python suite, and the full cross-platform CI matrix were not run locally.
- The unchanged original target tree has a known MSVC C2976 failure around `Array<Variant<int64_t, String>>` container subsumption; this history-only split does not fix it.
- The unchanged target has Doxygen warnings for new subsumption/`Optional` specializations and enum APIs; documentation cleanup remains follow-up work.

### CODE REVIEW

File: python/tvm_ffi/stub/python_generator/consts.py [hunk 1]
```diff
@@ -36,6 +36,7 @@
     "Tensor": "ffi.Tensor",
     "dtype": "ffi.dtype",
     "Device": "ffi.Device",
+    "ffi.Enum": "tvm_ffi.dataclasses.Enum",
```
Comment (junrushao) [inline python/tvm_ffi/stub/python_generator/consts.py:39]: TODO: remove this

File: tests/python/utils/filelock_worker.py [hunk 2]
```diff
@@ -15,31 +15,37 @@
 # specific language governing permissions and limitations
 # under the License.
 
+import os
```
Comment (junrushao) [inline tests/python/utils/filelock_worker.py]: TODO: revert changes in lock-related logics unless they are proven useful
